### ISSUE TITLE
Incorrect docstring / documentation for torch.nn.functional.scaled_dot_product_attention in 2.1

### ISSUE THREAD

**Issue by BroadbentJim** [pytorch/pytorch#110595]
### 📚 The doc issue

I originally noticed this error from following the documentation page at:
'https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html#torch.nn.functional.scaled_dot_product_attention'

where I noticed that the written 'efficient implementation' was not using the mask correctly:

```
# Efficient implementation equivalent to the following:
def scaled_dot_product_attention(query, key, value, attn_mask=None, dropout_p=0.0, is_causal=False, scale=None) -> torch.Tensor:
    # Efficient implementation equivalent to the following:
    L, S = query.size(-2), key.size(-2)
    scale_factor = 1 / math.sqrt(query.size(-1)) if scale is None else scale
    attn_bias = torch.zeros(L, S, dtype=query.dtype)
    if is_causal:
        assert attn_mask is None
        temp_mask = torch.ones(L, S, dtype=torch.bool).tril(diagonal=0)
        attn_bias.masked_fill_(temp_mask.logical_not(), float("-inf"))
        attn_bias.to(query.dtype)

    if attn_mask is not None:
        if attn_mask.dtype == torch.bool:
            attn_mask.masked_fill_(attn_mask.logical_not(), float("-inf"))
        else:
            attn_bias += attn_mask
    attn_weight = query @ key.transpose(-2, -1) * scale_factor
    attn_weight += attn_bias
    attn_weight = torch.softmax(attn_weight, dim=-1)
    attn_weight = torch.dropout(attn_weight, dropout_p, train=True)
    return attn_weight @ value
```

The error is that if the attn_mask.dtype == torch.bool, the attn_mask does not impact the attn_weight as it does not change the attn_bias and also is not mentioned in the final 5 lines. This documentation page is derived from the docstring in the code that has this error.

If we look at the current documentation page representing the docstring for the unstable version of the package at: https://pytorch.org/docs/main/generated/torch.nn.functional.scaled_dot_product_attention.html#torch.nn.functional.scaled_dot_product_attention, we can see:

```
 # Efficient implementation equivalent to the following:
def scaled_dot_product_attention(query, key, value, attn_mask=None, dropout_p=0.0, is_causal=False, scale=None) -> torch.Tensor:
    # Efficient implementation equivalent to the following:
    L, S = query.size(-2), key.size(-2)
    scale_factor = 1 / math.sqrt(query.size(-1)) if scale is None else scale
    attn_bias = torch.zeros(L, S, dtype=query.dtype)
    if is_causal:
        assert attn_mask is None
        temp_mask = torch.ones(L, S, dtype=torch.bool).tril(diagonal=0)
        attn_bias.masked_fill_(temp_mask.logical_not(), float("-inf"))
        attn_bias.to(query.dtype)

    if attn_mask is not None:
        if attn_mask.dtype == torch.bool:
            attn_bias.masked_fill_(attn_mask.logical_not(), float("-inf"))
        else:
            attn_bias += attn_mask
    attn_weight = query @ key.transpose(-2, -1) * scale_factor
    attn_weight += attn_bias
    attn_weight = torch.softmax(attn_weight, dim=-1)
    attn_weight = torch.dropout(attn_weight, dropout_p, train=True)
    return attn_weight @ value
```

where notably they masked_fill attn_bias instead of attn_mask.

N.B: I haven't really spent the time checking whether the code is implemented as in the docstring: READ: Incorrectly. 

### Suggest a potential alternative/fix

A fix is clear. Change the docstring that is used in V2.1 so as to have attn_bias.masked_fill_(attn_mask.logical_not(), float("-inf")) as in the unstable version
rather than attn_mask.masked_fill_(attn_mask.logical_not(), float("-inf")).

It may be worth considering whether this change can be explicitly made to the online documentation before doing a v2.1.1 where this error is amended.

cc @svekars @carljparker

**Comment by BroadbentJim** [comment]
Just a brief sanity check that in fact, the code is operating not as written but is operating as intended, by feeding a bo.

```
In [1]: import torch
   ...: import torch.nn.functional as F
   ...: print(torch.__version__)
   ...: # Optionally use the context manager to ensure one of the fused kernels is run
   ...: query = torch.rand(32, 8, 128, 64, dtype=torch.float16, device="cuda")
   ...: key = torch.rand(32, 8, 128, 64, dtype=torch.float16, device="cuda")
   ...: value = torch.rand(32, 8, 128, 64, dtype=torch.float16, device="cuda")
   ...: attn_mask = torch.randint(low=0,high=2,size= (32,8,128,128), device='cuda') > 0
   ...: torch.allclose(F.scaled_dot_product_attention(query,key,value, attn_mask=attn_mask), F.scaled_dot_product_atten
   ...: tion(query,key,value, attn_mask=None))
2.1.0+cu121
Out[1]: False

```

**Comment by jon-chuang** [comment]
This was fixed. https://github.com/pytorch/pytorch/pull/109086. 

Your script is wrong.

```python
import torch
import torch.nn.functional as F
print(torch.__version__)
# Optionally use the context manager to ensure one of the fused kernels is run
query = torch.rand(32, 8, 128, 64, dtype=torch.float16, device="cuda")
key = torch.rand(32, 8, 128, 64, dtype=torch.float16, device="cuda")
value = torch.rand(32, 8, 128, 64, dtype=torch.float16, device="cuda")
attn_mask_bool = torch.randint(low=0,high=2,size=(32,8,128,128), device='cuda') > 0
attn_mask_float = torch.zeros((32,8,128,128), device='cuda', dtype=torch.float16)
attn_mask_float.masked_fill_(attn_mask_bool.logical_not(), float("-inf"))

res = torch.allclose(
    F.scaled_dot_product_attention(query,key,value, attn_mask=attn_mask_bool),
    F.scaled_dot_product_attention(query,key,value, attn_mask=attn_mask_float)
)
print(res) # returns True
```

I think we should close this issue

**Comment by BroadbentJim** [comment]
The point of the issue was to draw attention to the fact that the incorrect docstring was in Pytorch 2.1, I hadn't seen the other issue that it had been resolved but I did point out that the correct docstring was in the latest build. I was coming at this from the PoV of someone having not used the function before and having the documentation say that it did not depend on attn_mask.

I think that you will find that the purpose of my script was to show that the function has some dependence on attn_mask in the boolean case  contradicting what the documentation says. I didn't really make it clear that that was the purpose of the little script, but I disagree that it is wrong.

**Comment by jainapurva** [comment]
Closing the issue as the documentation is fixed in both stable and main to use `attn_bias.masked_fill_(attn_mask.logical_not(), float("-inf"))`

`# Efficient implementation equivalent to the following:
def scaled_dot_product_attention(query, key, value, attn_mask=None, dropout_p=0.0, is_causal=False, scale=None) -> torch.Tensor:
    L, S = query.size(-2), key.size(-2)
    scale_factor = 1 / math.sqrt(query.size(-1)) if scale is None else scale
    attn_bias = torch.zeros(L, S, dtype=query.dtype)
    if is_causal:
        assert attn_mask is None
        temp_mask = torch.ones(L, S, dtype=torch.bool).tril(diagonal=0)
        attn_bias.masked_fill_(temp_mask.logical_not(), float("-inf"))
        attn_bias.to(query.dtype)

    if attn_mask is not None:
        if attn_mask.dtype == torch.bool:
            attn_bias.masked_fill_(attn_mask.logical_not(), float("-inf"))
        else:
            attn_bias += attn_mask
    attn_weight = query @ key.transpose(-2, -1) * scale_factor
    attn_weight += attn_bias
    attn_weight = torch.softmax(attn_weight, dim=-1)
    attn_weight = torch.dropout(attn_weight, dropout_p, train=True)
    return attn_weight @ value`
